docs: initial docs page for camunda-exporter (#4500)
camunda/camunda-docs: docs/self-managed/zeebe-deployment/exporters/camunda-exporter.md

---
id: camunda-exporter
title: "Camunda exporter"
sidebar_label: "Camunda Exporter"
description: "The common camunda exporter exporting zeebe records to harmonized web-apps indices ."
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

> The Camunda Exporter replaces the old Elasticsearch and Opensearch exporters. Previously, these exporters mainly

Author: I'm doubtful myself if `replaces the old Elasticsearch and Opensearch exporters` is a correct statement.

Author: I removed this statement and re-phrased the paragraph

---
id: camunda-exporter
title: "Camunda exporter"
sidebar_label: "Camunda Exporter"

> description: "The common camunda exporter exporting zeebe records to harmonized web-apps indices ."

Reviewer: ```suggestion
description: "The Camunda exporter exporting aggregated data to harmonized web-apps indices."
```

---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

The Camunda Exporter replaces the old Elasticsearch and Opensearch exporters. Previously, these exporters mainly
exported raw records from Zeebe streams into several indices. Camunda webapps (e.g., Operate, Tasklist) then imported
and transformed these records into useful forms. The Camunda Exporter aims to write records directly into
Elasticsearch/Opensearch in the form needed by the webapps, eliminating the need for additional importing or
transformation. This increases process instance throughput and reduces the latency of changes appearing
in Operate and Tasklist webapps by cutting off the importer from the data pipeline.

:::note
The indexes are created as required, and will not be created twice if they already exist. However, once disabled, they
will not be deleted (that is up to the administrator.) Similarly, data is never deleted by the exporter, and must be

> deleted by the administrator when it is safe to do so.

Reviewer: This is actually not true. Identity will delete users.

Author: that's true for users indeed, not for any other records though. I will think of a way to re-write this or remove it.

Author: I removed the statement completely for now

A [retention](#retention) policy can be configured to automatically delete data after a certain number of days.
:::

## Configuration

:::note
As the exporter is packaged with Zeebe, it is not necessary to specify a `jarPath`.

> :::

Reviewer: 🙃 I feel like these notes are more confusing than helpful. If I read this guide the first time I don;t know what a jarPath is.

Reviewer: Maybe we move it downwards

versions of Elasticsearch and/or Opensearch are supported in a Camunda 8 Self-Managed setup.
:::

| Option         | Description                                                                                                                   | Default                     |
| -------------- | ----------------------------------------------------------------------------------------------------------------------------- | --------------------------- |
| type           | the type of the underlying search engine to export to. Accepted values are `elasticsearch` or `opensearch`.                   | elasticsearch               |
| clusterName    | The name of the elasticsearch/opensearch cluster to export to.                                                                | elasticsearch               |
| dateFormat     | Defines a custom date format that should be used for fetching date data from the engine (should be the same as in the engine) | yyyy-MM-dd'T'HH:mm:ss.SSSZZ |
| socketTimeout  | Defines the socket timeout in milliseconds, which is the timeout for waiting for data.                                        |                             |

> | connectTimeout | Determines the timeout in milliseconds until a connection is established.                                                     |                             |

Reviewer: 🔧 I still find this confusing. I feel we need to change this somehow or make this simpler. It is not clear to me why we have these both.

Reviewer: What happens after the timeout?

Author: we can configure both separately in the client. we had one configuration variable before but actually we used this same value to set both variables in the client.

Also for operate importer they had both variables configured separately.

I don't see where does your confusion come from, but maybe if you elaborate a bit more, it could help me better understand.

>What happens after the timeout?

From the es/os client perspective, if it fails to establishes a connection before the configured `connectTimeout` it will throw a `ConnectionTimeoutException`, and if it waits longer than the `socketTimeout` to get a response to one of the client request, it will throw `SocketTimeoutException`. And both are actually separate concepts and, I believe, deserve separate configurations.

Reviewer: Maybe you can explain me (me as a user using Camunda 8) why I should configure them both and why different values would be beneficial for me.

Author: >Maybe you can explain me (me as a user using Camunda 8) why I should configure them both and why different values would be beneficial for me.

If you don't see a value as a user, then don't. just use the same value for both. if you know what you are doing, and have a good reason to have different values for those, then you have an option.

 >Thanks. What happens then ? Is the exporter retrying or are we stuck? Why should I configure a higher or lower value.

We create the client in the `Exporter.open()`, if it fails to connect, it will retry until successful.

We make requests to the exporter in subsequent exports or flushes, etc. if those fail due to a socket-timeout, the exporter `last exported position` should not change, and the next export will retry flushing the failed batches.

Reviewer: Thanks Mustafa.

| certificatePath | The file path to the SSL certificate used for secure communication with Elasticsearch/Opensearch. |         |
| verifyHostname  | If `true`, the hostname of the SSL certificate will be validated.                                 | true    |
| selfSigned      | If `true`, allows the use of self-signed SSL certificates.                                        | false   |

</TabItem>

<TabItem value="index">

| Option                | Description                                                                                                                                                                                 | Default |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- |
| prefix                | This prefix will be appended to every index created by the exporter; must not contain `_` (underscore).                                                                                     |         |

> | numberOfShards        | The number of [shards](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#_static_index_settings) used for each new record index created.                   | 3       |
> | numberOfReplicas      | The number of shard [replicas](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#dynamic-index-settings) used for each new record index created.           | 0       |

Reviewer: ```suggestion
| numberOfShards        | The number of [shards](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#_static_index_settings) used for each created index.                   | 3       |
| numberOfReplicas      | The number of shard [replicas](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#dynamic-index-settings) used for each created index.           | 0       |
```

| variableSizeThreshold | Defines a threshold for variable size. Variables exceeding this threshold are split into two properties: `FULL_VALUE` (full content, not indexed) and `VALUE` (truncated content, indexed). | 8191    |
| shardsByIndexName     | A map where the key is the index name and the value is the number of shards, allowing you to override the default `numberOfShards` setting for specific indices.                            |         |

> | replicasByIndexName   | A map where the key is the index name and the value is the number of replicas, allowing you to override the default `numberOfReplicas` setting for specific indices.                        |         |

Reviewer: ❓ This means setting this would also win right?

Author: yes, the ones in the map will override the default values set above

</TabItem>

<TabItem value="bulk">

To avoid too many expensive requests to the Elasticsearch/Opensearch cluster, the exporter performs batch
updates by default. The size of the batch, along with how often it should be flushed (regardless of
size) can be controlled by configuration.

| Option | Description                                                                                                                                                    | Default |
| ------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------- |
| delay  | Delay, in seconds, before force flush of the current batch. This ensures that even when we have low traffic of records, we still export every once in a while. | `5`     |
| size   | The amount of records a batch should have before we flush the batch                                                                                            | `1000`  |

With the default configuration, the exporter will aggregate records and flush them to Elasticsearch/Opensearch:

> 1. When it has aggregated 1000 records. 2. Five seconds have elapsed since the last flush (regardless of how many records were aggregated).

Reviewer: ```suggestion
1. When it has aggregated 1000 records. 
2. Five seconds have elapsed since the last flush (regardless of how many records were aggregated).
```

</TabItem>

<TabItem value="retention">

A retention policy can be set up to delete old data.
When enabled, this creates an Index Lifecycle Management (ILM) Policy that deletes the data after the specified
`minimumAge`.
All index templates created by this exporter apply the created ILM Policy.

| Option     | Description                                                                  | Default |
| ---------- | ---------------------------------------------------------------------------- | ------- |
| enabled    | If `true` the ILM Policy is created and applied to the index templates       | `false` |
| minimumAge | Specifies how old the data must be, before the data is deleted as a duration | `30d`   |

> | policyName | The name of the created and applied ILM policy                               |         |

Reviewer: ❓ There is a default value for the policy name right?

Author: we used to have a default in zeebe, but here not.

Reviewer: Isn;t this an issue then? If I enable this and have no name it would fail no? I need a default name I think

Author: Not really an issue, if you enable retention, you should give a name to the policy. it's not consistent with other exporters, but not an issue.

I can't remember if that was a decision, but I guess it was removed when there was a confusion whether we will have different policies or not.

Anyway, we can bring back a default value, but here I'm just documenting what there is right now.

:::note
The duration can be specified in days `d`, hours `h`, minutes `m`, seconds `s`, milliseconds `ms`, and/or nanoseconds
`nanos`.
:::

</TabItem>
</Tabs>

## Example

Here is an example configuration of the exporter:

```yaml
---
exporters:
  # Camunda Exporter ----------
  # An example configuration for the camunda exporter:
  #

>   # These setting can also be overridden using the environment variables "ZEEBE_BROKER_EXPORTERS_CAMUNDAEXPORTER_..."

Reviewer: 🔧 Worth to mention how the args look like, because this often lead to confusion, as this is not with `_`

Author: can youn explain a bit more what do you mean?

Reviewer: Was thinking about an example of setting something in the ARGS section like:

`ZEEBE_BROKER_EXPORTERS_CAMUNDAEXPORTER_ARGS_INDEX_NUMBEROFSHARDS`

For example related support case. https://jira.camunda.com/browse/SUPPORT-23300

Author: I added a note on how to convert it and an example below

Reviewer: Double underscore?

Author: oh no, that's wrong. I guess I relied on co-pilot completing the sentence and didn't check for correctness 😄

Author: fixed

---
id: camunda-exporter
title: "Camunda exporter"
sidebar_label: "Camunda Exporter"
description: "The Camunda exporter exporting aggregated data to harmonized web-apps indices."
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

The Camunda Exporter is designed to export Zeebe records directly into Elasticsearch/Opensearch in the form needed by
Camunda webapps (e.g., Operate, Tasklist). Previously, the Elasticsearch and Opensearch exporters mainly exported raw
records from Zeebe streams into several indices, which were then imported and transformed by the webapps. The Camunda
Exporter eliminates the need for additional importing or transformation, increasing process instance throughput and
reducing the latency of changes appearing in Operate and Tasklist webapps by cutting off the importer from the data
pipeline.

:::note
The indexes are created as required, and will not be created twice if they already exist. However, once disabled, they
will not be deleted (that is up to the administrator.)
A [retention](#retention) policy can be configured to automatically delete data after a certain number of days.
:::

## Configuration

> The exporter can be enabled by configuring it with the `classpath` in the broker settings.

Reviewer: ```suggestion
The exporter can be enabled by configuring it with the `className` in the broker settings.
```
